### ISSUE TITLE
Window#focus() brings all app windows to front

### ISSUE THREAD

**Issue by szhu** [kasper/phoenix#148]
- Version: 2.3
- macOS: 10.11.6

When working on multiple displays, focusing a window brings that app's other windows to the front as well.

| Demo of the bug: |
|:----------------- |
| The bottom Chrome window should not have risen above the Sublime Text window, but it does. |
| ![output_spdww3](https://cloud.githubusercontent.com/assets/1570168/19944799/77422052-a0fa-11e6-8fa8-8215aac0d952.gif) |

Possibly useful observation: I see the bottom Chrome window quickly flash in and out of focus immediately before the top Chrome window comes into focus. This happens too quickly to be screen-captured.

Phoenix code used (this focuses the window under the mouse)
```coffee
#!/usr/bin/env coffee -p

within = (point, rect) ->
  return Math.min(
    point.x - rect.x,
    point.y - rect.y,
    rect.x + rect.width - point.x,
    rect.y + rect.height - point.y,
  )

Event.on 'mouseDidMove', ->
  mouseLocation = Mouse.location()
  for window in Window.recent()
    if within(mouseLocation, window.frame()) > 0
      window.focus()
      break
```

JavaScript version:

```js
function within(point, rect) {
  return Math.min(
    point.x - rect.x,
    point.y - rect.y,
    rect.x + rect.width - point.x,
    rect.y + rect.height - point.y
  );
};

Event.on('mouseDidMove', function() {
  var mouseLocation = Mouse.location();
  var windows = Window.recent();
  for (var i = 0; i < windows.length; i++) {
    if (within(mouseLocation, windows[i].frame()) > 0) {
      windows[i].focus();
      break;
    }
  }
});
```

**Comment by mafredri** [comment]
I believe this is a Chrome issue. I've tried dealing with it before but unfortunately there doesn't seem to be a way around it.

Although I can't reproduce this any more, maybe it's because I stopped using Chrome Apps, are you using any perhaps?

**Comment by szhu** [comment]
I'm not using any Chrome apps. But ah, you're totally right, this is a Chrome issue. I can reproduce this using Spark, a shortcut manager. No other apps I tried exhibit this behavior.

**Comment by kasper** [comment]
Yes, unfortunately Chrome seems to have some weird window management behaviours.

**Comment by szhu** [comment]
What's the equivalent Obj-C API call that is buggy in Chrome? I can try to file a bug for this.

**Comment by szhu** [comment]
Actually, nope I can reproduce this issue with most other apps.

**Comment by kasper** [comment]
`Window#focus()` needs to bring the app focused, before it can set a window to be its main window. If it doesn’t do that, it will most likely fail. That probably is causing the difference in some apps. I can see if this can be tweaked somehow.

**Comment by szhu** [comment]
Does this look promising?
https://developer.apple.com/reference/appkit/nsaccessibilityraiseaction

**Comment by kasper** [comment]
@szhu Definitely, sounds promising! I’ll have to test it out. Thanks! :simple_smile:

**Comment by kasper** [comment]
I can’t reproduce this behaviour on macOS 10.12.1, Phoenix 2.3 and Chrome 54.0.2840.98. Only the actual window comes front.

```javascript
Key.on('r', [ 'ctrl', 'shift' ], function () {
  App.get('Google Chrome').windows()[0].focus();
});
```

**Comment by mafredri** [comment]
@kasper I think you need to have multiple Chrome windows and focus another window than the last/previously focused window (of Chrome). Perhaps even multiple monitors is required, not sure.

**Comment by kasper** [comment]
@mafredri Two monitors, multiple Chrome windows, focusing another Chrome window than the previously focused window — still Phoenix only focuses the window I’m expecting it to focus.

**Comment by mafredri** [comment]
Ok, then I don't know. I've not been able to reproduce either, either due
to updated Chrome or Sierra (I run Chrome beta btw).

**Comment by kasper** [comment]
I’ll close this because I cannot reproduce this nor do I know if I could fix it in Phoenix. I’ll reopen if we figure out what the cause is.

I implemented the raise action in #151 which raises the window so it will be beneath the focused window. (The documented description seems to differ slightly.) Most likely won’t help here though.
